### PR TITLE
Samd51 flash storage compatibility

### PR DESCRIPTION
**Pull request by OnlineCaveman**
I wanted to use the repository on a SAMD51 and noticed the flash memory routines were only for SAMD21 so here are my proposed changes for SAMD51 compatibility.

Since I have no SAMD21 board available at the moment, I can only verify the code on a SAMD51 platform.

### CODE REVIEW

**Review by SirSydom — COMMENTED** [review]
maybe in that case a samd51_plattform - derived from the samd_plattform would be a better solution then extensive ifdef usage...

**Comment by OnlineCaveman** [comment]
Good suggestion. For now my time is limited. I will change this to a draft and come back to it.

File: src/knx/knx_value.cpp [hunk 1]
```diff
@@ -1,562 +1,562 @@
...
+{
+    _value.doubleValue = value;
+    _type = DoubleType;
+    return *this;
+}
+
+KNXValue::operator float() const
+{
+    return doubleValue();
 }
```
Comment (thelsing) [inline src/knx/knx_value.cpp:562]: Did you change anything here? Where the line endings wrong?

File: src/samd_platform.cpp [hunk 2]
```diff
@@ -1,6 +1,10 @@
+#if(defined(__SAMD21E17A__) || \
+    defined(__SAMD21G18A__) || \
+    defined(__SAMD21E18A__) || \
+    defined(__SAMD21J18A__))
+
 #include "samd_platform.h"
 
-#ifdef ARDUINO_ARCH_SAMD
```
Comment (thelsing) [inline src/samd_platform.cpp:3]: Any reason to not also put
 #if (defined(__SAMD21E17A__) || \
     defined(__SAMD21G18A__) || \
     defined(__SAMD21E18A__) || \
     defined(__SAMD21J18A__)) 
here?

File: src/knx/knx_value.cpp (hunk 1 above)
Comment (OnlineCaveman) [reply src/knx/knx_value.cpp:562]: No, nothing changed as far as I know. Was surprised to see it being included in the push.

File: src/samd_platform.cpp (hunk 2 above)
Comment (OnlineCaveman) [reply src/samd_platform.cpp:3]: No, copy paste ended up at the start of the file. I will move them after the include samd_platform.h to be consistent with the other files.

**Comment by OnlineCaveman** [comment]
Base work for W5500 support made. Errors in macro expansion getbyte

Intellisense message:
#define htonl(x) ( (getbyte(x,0)<<24) | (getbyte(x,1)<<16) | (getbyte(x,2)<<8) | getbyte(x,3) )
Expands to:
( (getbyte(io->_platform.currentIpAddress(),0)<<24) | (getbyte(io->_platform.currentIpAddress(),1)<<16) | (getbyte(io->_platform.currentIpAddress(),2)<<8) | getbyte(io->_platform.currentIpAddress(),3) )
expression must be an lvalue or a function designatorC/C++(158)

Build error:
In file included from src\knx\ip_parameter_object.cpp:5:
src\knx\ip_parameter_object.cpp: In lambda function:
src\knx\bits.h:9:39: error: lvalue required as unary '&' operand
    9 | #define getbyte(x,n) (*(((uint8_t*)&(x))+n))
      |                                       ^
src\knx\bits.h:11:21: note: in expansion of macro 'getbyte'
   11 | #define htonl(x) ( (getbyte(x,0)<<24) | (getbyte(x,1)<<16) | (getbyte(x,2)<<8) | getbyte(x,3) )
      |                     ^~~~~~~
src\knx\ip_parameter_object.cpp:49:25: note: in expansion of macro 'htonl'
   49 |                 pushInt(htonl(io->_platform.currentIpAddress()), data);
      |                         ^~~~~

Anyone a clue as why this is the case? The samd51platform::currentIpAddress() returns Ethernet.localIP() analogue to WiFi.localIP() as used in the esp_platform and esp32_platform cases, both of which compile fine.

**Comment by thelsing** [comment]
Maybe we should put ARCH defines around htonl and friends so that we only provide them if the platform doesn't have them already.

**Review by thelsing — CHANGES_REQUESTED** [review]
Please get rid of the changes of knx_value or at least investigate what  changed. (If this was the only file with different line endings this change would be ok)

In knx.demo is too much stuff now. The setup code for Ethernet should be moved to the Platform. The esp-Platforms have examples on how to do this.

Also please don't remove the samd-zero env from platform.ini. You are free to add the new env  for adafruit_feather_m4.

**Comment by thelsing** [comment]
I just merged this manually into devel. Thanks for your contribution.

**Pull request closed without merging**